nni/retiarii/oneshot/pytorch/differentiable.py

```diff
@@ -0,0 +1,354 @@
+                continue
+            yield name, p
+
+    def export(self):
+        return list(self.op_choices.keys())[torch.argmax(self.alpha).item()]
+    
+    def r_forward(self, x):
+        return self.op_choices.values()[0](x)
+
+class DartsInputChoice(nn.Module):
+    def __init__(self, input_choice):
+        super(DartsInputChoice, self).__init__()
+        self.name = input_choice.label
+        self.alpha = nn.Parameter(torch.randn(input_choice.n_candidates) * 1e-3)
+        self.n_chosen = input_choice.n_chosen or 1
+
+    def forward(self, inputs):
+        inputs = torch.stack(inputs)
+        alpha_shape = [-1] + [1] * (len(inputs.size()) - 1)
+        return torch.sum(inputs * F.softmax(self.alpha, -1).view(*alpha_shape), 0)
+
+    def parameters(self):
+        for _, p in self.named_parameters():
+            yield p
+
+    def named_parameters(self, recurse):
+        for name, p in super(DartsInputChoice, self).named_parameters():
+            if name == 'alpha':
+                continue
+            yield name, p
+
+    def export(self):
+        return torch.argsort(-self.alpha).cpu().numpy().tolist()[:self.n_chosen]
+    
+    def r_forward(self, x):
+        return super().r_forward(x)
+
+DARTS_REPLACE_DICT = {
+    LayerChoice : DartsLayerChoice,
+    InputChoice : DartsInputChoice
+}
+
+class DartsModel(BaseOneShotLightningModule):
+    '''
+    The DARTS model. In each iteration, there are 2 training phases. The phase 1 is architecture
+    step, in which the model parameters are frozen and the alphas are optimized. The phase 2 is model
+    step, in wchich the alphas are frozen and the model parameters are optimized. In this step, the
+    out put of each choice are sumed up with respect to their alpha value. The result of DARTS
+    is argmax in alpha.
+    The DARTS Model should be trained with ParallelTraiValDataloader in nn.retiarii.oneshot.pytorch.utils.
+    See base class for more attributes.
+
+    Reference
+    ----------
+    [1]H. Liu, K. Simonyan, and Y. Yang, “DARTS: Differentiable Architecture Search,” presented at the
+    International Conference on Learning Representations, Sep. 2018. Available: https://openreview.net/forum?id=S1eYHoC5FX
+    '''
+    def __init__(self, model, custom_replace_dict = None):
+        '''
+        Parameters
+        ----------
+        base_model : pl.LightningModule
+            The module in evaluators in nni.retiarii.evaluator.lightning. User defined model
+            is wrapped by base_model, and base_model will be wrapped by this model.
+        custom_replace_dict : dict{ type : callable }, default = None
+            The custom xxxChoice replace method. Keys should be xxxChoice type and values should 
+            return an nn.module. This custom replace dict will override the default replace
+            dict of each nas method.
+        '''
+        super().__init__(model, custom_replace_dict)
+        self.automatic_optimization = False
```

Comment:
Put this outside and says `automatic_optimization = False`.

Reply:
The lightning docs uses self.automatic_optimization = False https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#manual-optimization.  Shall we follow their style?

Reply:
I think it's not necessary.